qiskit/algorithms/optimizers/spsa.py

```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
+                              np.linalg.norm(update),  # size of the update step
+                              True)  # accepted
+
+            # update parameters
+            x = x_next
+
+            # update the list of the last ``last_avg`` parameters
+            if self.last_avg > 1:
+                last_steps.append(x_next)
+                if len(last_steps) > self.last_avg:
+                    last_steps.popleft()
+
+            # check termination
+            if np.linalg.norm(update) < self.tolerance:
+                break
+
+        logger.info('SPSA finished in %s', time() - start)
+        logger.info('=' * 30)
+
+        if self.last_avg > 1:
+            x = np.mean(last_steps, axis=0)
+
+        return x, loss(x), self._nfev
 
     def get_support_level(self):
-        """ return support level dictionary """
+        """Get the support level dictionary."""
         return {
-            'gradient': OptimizerSupportLevel.ignored,
+            'gradient': OptimizerSupportLevel.ignored,  # could be supported though
             'bounds': OptimizerSupportLevel.ignored,
             'initial_point': OptimizerSupportLevel.required
         }
 
     def optimize(self, num_vars, objective_function, gradient_function=None,
                  variable_bounds=None, initial_point=None):
-        super().optimize(num_vars, objective_function, gradient_function,
-                         variable_bounds, initial_point)
+        return self._minimize(objective_function, initial_point)
 
-        if not isinstance(initial_point, np.ndarray):
-            initial_point = np.asarray(initial_point)
 
-        logger.debug('Parameters: %s', self._parameters)
-        if not self._skip_calibration:
-            # at least one calibration, at most 25 calibrations
-            num_steps_calibration = min(25, max(1, self._maxiter // 5))
-            self._calibration(objective_function, initial_point, num_steps_calibration)
-        else:
-            logger.debug('Skipping calibration, parameters used as provided.')
-
-        opt, sol, _, _, _, _ = self._optimization(objective_function,
-                                                  initial_point,
-                                                  maxiter=self._maxiter,
-                                                  **self._options)
-        return sol, opt, None
-
-    def _optimization(self,
-                      obj_fun: Callable,
-                      initial_theta: np.ndarray,
-                      maxiter: int,
-                      save_steps: int = 1,
-                      last_avg: int = 1) -> List:
-        """Minimizes obj_fun(theta) with a simultaneous perturbation stochastic
-        approximation algorithm.
+def bernoulli_perturbation(dim, perturbation_dims=None):
+    """Get a Bernoulli random perturbation."""
+    if perturbation_dims is None:
+        return np.array([1 - 2 * np.random.binomial(1, 0.5) for _ in range(dim)])
```

Comment:
why does this not use the `size` kwarg in the `binomial` call?

Reply:
it should, thanks!